### PR TITLE
ARROW-12812: [Packaging][Java] Improve JNI jars build

### PR DESCRIPTION
**Pull request by kszucs**

### CODE REVIEW

File: ci/scripts/java_jni_build.sh [hunk 1]
```diff
@@ -21,21 +21,20 @@ set -e
...
-  fi
+# build the entire project
+mvn clean install -DskipTests -P arrow-jni -Darrow.cpp.build.dir=$cpp_build_dir
+# test jars that have cpp dependencies
+mvn test -P arrow-jni -pl adapter/orc,gandiva,dataset -Dgandiva.cpp.build.dir=$cpp_build_dir
+
+# copy the jars that has cpp dependencies to distribution folder
+find gandiva/target/ -name "*.jar" -not -name "*tests*" -exec cp {} $cpp_build_dir \;
+find adapter/orc/target/ -name "*.jar" -not -name "*tests*" -exec cp {} $cpp_build_dir \;
+find dataset/target/ -name "*.jar" -not -name "*tests*" -exec cp {} $cpp_build_dir \;
```
Comment (anthonylouisbsb) [inline ci/scripts/java_jni_build.sh:38]: I think you can modify this block to generate the jars for all modules, my suggestion:

```shell  
# build and test the entire project
  mvn clean install -P arrow-jni -Darrow.cpp.build.dir=$cpp_build_dir

  MODULES=(
    gandiva adapter/orc adapter/avro adapter/jdbc dataset algorithm
    compression format performance tools vector plasma flight/flight-core
    flight/flight-grpc memory/memory-core memory/memory-netty memory/memory-unsafe
  )

  # copy all jars to distribution folder, excluding the unit tests ones
  for module in "${MODULES[@]}"
  do
    find $module/target/ -name "*.jar" -not -name "*tests*" -not -name "*benchmarks*" -exec cp  {} $cpp_build_dir \;
  done
```

**Review by anthonylouisbsb — COMMENTED** [review]
@kszucs I was working to update the script [to generate the jars to all modules](https://issues.apache.org/jira/browse/ARROW-12801). But I noticed that you made some changes that solve the problem, so I made a suggestion to you apply in java build script to generate jars for all modules and you should apply [these changes](https://github.com/s1mbi0se/arrow/blob/7316dc3b8cf38f73cbd1e050167bb96b2d180ceb/dev/tasks/tasks.yml#L628) to update the list to uploaded artifacts by the task.

Thanks in advance for your help and give me feedback if my suggestions are fine.

**Comment by kszucs** [comment]
> Can we also generate .jar that doesn't need to bundle C++ libraries as the next step? Then we don't need to generate any .jar in our release process.

@kou I was thinking of the exact same thing. Using Anthony's snippet above we can generate the jars, though I'm not sure how to upload them without recompiling the jars.

**Comment by kou** [comment]
@andygrove told that we can use `mvn deploy:deploy-file`: https://github.com/apache/arrow/pull/9155#issuecomment-820826684

**Comment by kszucs** [comment]
> @andygrove told that we can use `mvn deploy:deploy-file`: [#9155 (comment)](https://github.com/apache/arrow/pull/9155#issuecomment-820826684)

Great!

**Comment by kszucs** [comment]
@github-actions crossbow submit java-jars

**Comment by kszucs** [comment]
@github-actions crossbow submit java-jars

**Comment by kszucs** [comment]
The produced artifacts are here: https://github.com/ursacomputing/crossbow/releases/tag/actions-415-github-java-jars

**Comment by kszucs** [comment]
@github-actions crossbow submit java-jars

**Comment by kszucs** [comment]
@github-actions crossbow submit java-jars

**Comment by kszucs** [comment]
@github-actions crossbow submit java-jars

**Review by kszucs — COMMENTED** [review]
+1, the CI failures are unrelated.

**Pull request closed without merging**